### PR TITLE
Kappy dev

### PR DESCRIPTION
**Pull request by pagreene**
Improve the kappy python package.

### CODE REVIEW

**Comment by pagreene** [comment]
I had to fix some issues in the tests which I somehow missed. The tests now run and pass (on my machine, with both python 2.7 and 3.5). I have no idea what the issue is with the travis build, but it doesn't seem to have to do with my code.

**Comment by pirbo** [comment]
Thank you very very much for this work. I wish I had time to have a look today but I was completely under the water. I hope I will tomorrow.
The Travis failure of yesterday was indeed not your fault (I did something that broke the build of any pull request) but I fixed it. The failure of today on the contrary is a consequence of your changes:
Stripping out the other tests, the way `.travis.yml` is set up to test the python client is
```
language: python
python: 3.5
addons:
  apt:
    sources:
      - avsm
    packages:
      - ocaml-nox
      - aspcud
      - opam
install:
  - opam init -a git://github.com/ocaml/opam-repository
  - eval $(opam config env)
  - opam install -y atdgen cohttp ; fi"
  - pip2 install --user . nose && pip install -v . nose
script
  - make agents bin/WebSim
  - cd python && python3 tests/test_kappa_client.py && python2 tests/test_kappa_client.py
```
which obviously require some updates as you removed `test_kappa_client.py` (and that's good).

Do you know what would be the correct way to run both python2.7 nose tests and python3.5 one in a row?

**Comment by pagreene** [comment]
Ah yes, I'd forgotten about the yml file. I implemented a somewhat crude fix for now. It's not an elegant solution, but it should work.

**Comment by pagreene** [comment]
With indra, we actually have two separate builds, one each for 2.7 and 3.5. I don't think that paradigm would work here, though.

**Comment by pirbo** [comment]
There is an issue with where kappy looks for binaires once installed.
They are all looked in `/home/travis/virtualenv/python3.5.3/lib/python3.5/bin/` whereas
agents (KaSimAgent and KaSaAgent) are installed by pip in `/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/kappy/bin` and WebSim is not installed anywhere, the tests should look for it in the source directory.

**Comment by pagreene** [comment]
Well WebSim _must_ be installed, or else the tests can't run, but it looks like it's created in `../../bin/WebSim`, relative to the test directory, which would be correct except that I changed `KASIM_DIR` to be relative to the location of the kappa_common.py file. I hadn't realized that the travis tests were run using the pip install of kappy...that probably won't work at the moment anyway because I don't think the pip kappy is up to date with this branch (I hope not...).

As a matter of fact, I don't see where pip install kappy...so I'm really not sure why this doesn't work.

Anyways, the system should probably have a two-step system, where it first looks for an environment variable for the KaSim dir, then tries to find it in some default location (and actually verifiy that it's the KaSim directory), and then give up. Similarly, when looking for WebSim, it should look for an environment variable, and then in the default location....in principal, if we know where the KaSim directory is we could potentially have the test build WebSim if it doesn't find it, but that might be a bit too aggressive for the test.

**Comment by pirbo** [comment]
I haven't notice earlier but it looks like python2 tests fail immediately with
```
  File "tests/test_rest.py", line 2, in <module>
    from builtins import dict, str
ImportError: No module named builtins
```

**Comment by pagreene** [comment]
The point of the default sim parameter is (a) it is one less special class that would need to be imported from kappy to perform a simple task, and (b) it makes it easier when running the same simulation multiple times. And (c) it adds these options without interfering with the previous use-case. It's also not complicated or difficult to maintain.

**Comment by pagreene** [comment]
Ah, yes, that is a missing dependency that you don't have, which is needed for python 2-3 compatibility. Let me see if I can find the package that has that in it...

**Comment by pirbo** [comment]
`kappy` is "installed" by pip through `pip install -v .` so it "installs" the very last version of files.

Yes, WebSim is the easy part.
The tests should not import `KASIM_DIR` from KappaCommon and should simply look for WebSim relatively to themselves in `../../bin/WebSim`.

The hard part is to make KappaStd find the agents in both installed and non installed settings...
In the first case they are in `KAPPY_DIR/bin/` which is `KappaCommon.py_dir/bin` whereas in the second one they are indeed in `KASIM_DIR/bin/` which is `KappaCommon.py_dir/../../bin/`.

**Comment by pagreene** [comment]
Well it wouldn't be that hard to simply try both location and use whichever works. I wouldn't say the tests should _always_ look for WebSim relative to themselves. Is WebSim only used for these tests?

As for the missing dependency, I think I've pushed a commit that I hope will solve that.

**Comment by pirbo** [comment]
> Is WebSim only used for these tests?

Somehow, I would say yes.
The tests are the only one who has to launch its own web server to check that it is compatible with it.

I envision the normal use of KappaRest as being to reach an already existing external REST kappa server than launching your own private one locally on your machine. Why not using KappaStd in this case?

**Comment by pagreene** [comment]
But would said exiting REST server have been started using the WebSim command? And if so, wouldn't that command be stored in KASIM_DIR/bin?

I've pushed a change that should fix kappa_std.

**Comment by pagreene** [comment]
Alright, I think that may have fixed the issues. I had the system actually search for WebSim somewhere bellow one of the two directories. It's a fairly dumb search, in that it makes virtually no assumptions, and it could potentially take some time, but in general it shouldn't. The code could be easily generalized to work with finding the KaSa and KaSim agents as well, though that doesn't seem necessary at the moment.

File: python/tests/util.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+from builtins import dict, str
+
+import sys
+import unittest
+from os import path
+
+import kappy
+from kappy.kappa_common import KASIM_DIR
+
+MODELS_DIR = path.join(KASIM_DIR, "models")
```
Comment (pirbo) [inline python/tests/util.py]: Models are not installed either. So KASIM_DIR is not the good path is kappy is installed

**Comment by pagreene** [comment]
Well, that still didn't quite work. Now it can't find the test model files, and apparently WebSim couldn't be found anywhere...which is probably a glitch on my end.

**Comment by pagreene** [comment]
I think to fix the WebSim issue, we should just have it build in the test directory, and just do that in general.

As for the model files...don't we need those?

**Comment by pirbo** [comment]
by the way, it would be nice if failing tests could make the travis build fails. It looks like there are 2 solutions:
- add something like `sys.exit(not result.wasSuccessful())` if I understand https://stackoverflow.com/questions/24972098/unit-test-script-returns-exit-code-0-even-if-tests-fail correctly
- Using the `nosetests2` and `nosetests3` that seems to be available on ubuntu python installation... (I'm sure about the `nosetests3` one, I can't understand whether there is a `nosetests2` one or if you have to use `nosetests`(in which case, is that sure that it is always python2... ?))

**Comment by pirbo** [comment]
We need models to do some tests but no they are not part of the library, users come with their own models...

**Comment by pagreene** [comment]
Failed test = Travis fail: Yes, I was actually going to comment on that once I had a chance to research it (apparently you beat me to that :) ). I will try the nostests2/3 solution first, and we'll see what happens.

Yes, but the model files I use come from the git repository, I didn't make those. They are also the model files that were used by the tests before, although I believe those tests may have been commented out. In any case, I'm still not clear why the files in the git repo don't show up anywhere...or is it simply that there is yet another location where those have ended up that isn't KAPPY_DIR or KASIM_DIR.

**Comment by pagreene** [comment]
OK, actually, a third, cleaner solution is to have an extra build version that is just for python 2.7. That's what we do for indra, and I think it would make things both clearer, and simpler.

**Comment by pagreene** [comment]
So, it seems trying to have the separate python builds the way I did doesn't work...I was imitating what we do for indra, but at a lower level, so I'm not entirely surprised that didn't work. Unfortunately, that means we can't be sure if my new file finding scheme works or not.

**Comment by pagreene** [comment]
Well, it seems to be making some improvements, but now there seems to be an issue with setup.py...it looks vaguely like it is failing to build parts of KaSim...any insights?

**Comment by pirbo** [comment]
Bash is a nightmare :-). The problem is hidden in
```
sh: 1: [python: not found
```

you must replace
```
sh -c "if [ '$CHECKED' = webserver ] || ['$CHECKED' = python ] ; then opam install -y atdgen cohttp ; fi"
```
by
```
sh -c "if [ '$CHECKED' = webserver ] || [ '$CHECKED' = python ] ; then opam install -y atdgen cohttp ; fi"
```

**Comment by pirbo** [comment]
(the relevant line in the pip error was `make: atdgen: Command not found`)

By the way, the `CHECKED=python` tests are super sets of `CHECKED=webserver` one, you can simply remove it

**Comment by pagreene** [comment]
Which would I remove? The webserver test?

**Comment by pirbo** [comment]
Yes, the matrix "line" of `env: CHECKED=webserver`

**Comment by pirbo** [comment]
Thank you very much

**Pull request merged**